### PR TITLE
Add typed versions of Mirror.ProductOf#fromProduct

### PR DESCRIPTION
**Pull request by mpilquist**
The existing `Mirror.ProductOf#fromProduct` method is defined for any `scala.Product` which can lead to some surprising runtime errors:

```scala
Welcome to Scala 3.1.2-RC1-bin-SNAPSHOT-nonbootstrapped-git-b636aa9 (17.0.1, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.

scala> case class Point(x: Int, y: Int, z: Int)
// defined case class Point

scala> summon[deriving.Mirror.ProductOf[Point]].fromProduct((1, 2))
java.lang.IndexOutOfBoundsException: 2 is out of bounds (min 0, max 1)
  at scala.Product2.productElement(Product2.scala:43)
  at scala.Product2.productElement$(Product2.scala:40)
  at scala.Tuple2.productElement(Tuple2.scala:24)
  at rs$line$1$Point$.fromProduct(rs$line$1:1)
  at rs$line$1$Point$.fromProduct(rs$line$1:1)
  ... 38 elided

scala> summon[deriving.Mirror.ProductOf[Point]].fromProduct((1, 2, 3.0))
java.lang.ClassCastException: class java.lang.Double cannot be cast to class java.lang.Integer (java.lang.Double and java.lang.Integer are in module java.base of loader 'bootstrap')
  at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:99)
  at rs$line$1$Point$.fromProduct(rs$line$1:1)
  at rs$line$1$Point$.fromProduct(rs$line$1:1)
  ... 38 elided
```

This PR adds `fromTuple` and `fromProductTyped` extension methods to `Mirror.ProductOf[T]`, both which ensure the argument is well typed:

```scala
scala> summon[deriving.Mirror.ProductOf[Point]].fromTuple((1, 2))
-- [E007] Type Mismatch Error: -------------------------------------------------
1 |summon[deriving.Mirror.ProductOf[Point]].fromTuple((1, 2))
  |                                                   ^^^^^^
  |                                                 Found:    (Int, Int)
  |                                                 Required: (Int, Int, Int)

longer explanation available when compiling with `-explain`
1 error found

scala> summon[deriving.Mirror.ProductOf[Point]].fromTuple((1, 2, 3.0))
-- [E007] Type Mismatch Error: -------------------------------------------------
1 |summon[deriving.Mirror.ProductOf[Point]].fromTuple((1, 2, 3.0))
  |                                                          ^^^
  |                                                 Found:    (3.0d : Double)
  |                                                 Required: Int

longer explanation available when compiling with `-explain`
1 error found

scala> summon[deriving.Mirror.ProductOf[Point]].fromTuple((1, 2, 3))
val res2: Point = Point(1,2,3)
```

### CODE REVIEW

File: tests/neg/deriving.scala [hunk 1]
```diff
@@ -1,6 +1,6 @@
-import reflect.Generic
+import deriving.MirrorOf
 
-sealed trait A derives Generic // error: cannot take shape, it has anonymous or inaccessible subclasses
```
Comment (mpilquist) [inline tests/neg/deriving.scala:3]: These `Generic` tests appeared to date back to a much older version of `Mirror`. Adjusted for current state, though I could do this in a separate PR if desired.

**Comment by dwijnand** [comment]
Add as experimental?

**Comment by mpilquist** [comment]
> Add as experimental?

Done. Curious when `@experimental` gets removed from API additions? Any link to docs on that?

**Comment by nicolasstucki** [comment]
> Curious when @experimental gets removed from API additions?

We can only remove experimental when going to a new minor release (3.2.0, 3.3.0, 3.4.0, ...). Before each minor release we need to decide which experimental method will become non-experimental.

File: tests/neg/deriving.scala (hunk 1 above)
Comment (nicolasstucki) [reply tests/neg/deriving.scala:3]: If this still emits an error we should keep it as it is.

Comment (nicolasstucki) [reply tests/neg/deriving.scala:3]: It would be better to keep this file untouched and add the new version in a `tests/neg/deriving-b.scala` test file.

File: library/src/scala/deriving/Mirror.scala [hunk 2]
```diff
@@ -49,4 +49,15 @@ object Mirror {
   type Of[T] = Mirror { type MirroredType = T; type MirroredMonoType = T ; type MirroredElemTypes <: Tuple }
   type ProductOf[T] = Mirror.Product { type MirroredType = T; type MirroredMonoType = T ; type MirroredElemTypes <: Tuple }
   type SumOf[T] = Mirror.Sum { type MirroredType = T; type MirroredMonoType = T; type MirroredElemTypes <: Tuple }
+
+  extension [T](p: ProductOf[T])
+    /** Create a new instance of type `T` with elements taken from product `a`. */
+    @annotation.experimental
+    def fromProductTyped[A <: scala.Product](a: A)(using m: ProductOf[A], ev: p.MirroredElemTypes =:= m.MirroredElemTypes): T =
```
Comment (nicolasstucki) [inline library/src/scala/deriving/Mirror.scala:56]: I'm wondering if it is possible to do something like
```scala
    def fromProductTyped[A <: scala.Product](a: A)(using m: ProductOf[A] { type MirroredElemTypes = p.MirroredElemTypes }): T =

```

Comment (nicolasstucki) [reply library/src/scala/deriving/Mirror.scala:56]: It also feels like `p.MirroredElemTypes =:= m.MirroredElemTypes` should be a `<:<`. As it is safe to construct a `(List[Int], List[String])` from `(::[Int], ::[String])`.

We should have some positive and negative tests for these situations.

File: library/src/scala/deriving/Mirror.scala [hunk 3]
```diff
@@ -49,4 +49,15 @@ object Mirror {
...
+      p.fromProduct(a)
+
+    /** Create a new instance of type `T` with elements taken from tuple `t`. */
+    @annotation.experimental
+    def fromTuple(t: p.MirroredElemTypes): T =
```
Comment (nicolasstucki) [inline library/src/scala/deriving/Mirror.scala:61]: If `MirroredElemTypes` would be defined in `Mirror.Product` we could have defined this method directly in `Mirror.Product`.

@odersky was there a reason to put it where it is now? 
It seems that we could move it without breaking binary compatibility. Could we move it?

File: library/src/scala/deriving/Mirror.scala (hunk 2 above)
Comment (mpilquist) [reply library/src/scala/deriving/Mirror.scala:56]: I didn't use a refinement because of https://github.com/lampepfl/dotty/issues/7380 but just noticed that was fixed. Will give it a shot.

File: tests/neg/deriving.scala (hunk 1 above)
Comment (mpilquist) [reply tests/neg/deriving.scala:3]: OK I'll restore it. I thought, perhaps erroneously, that these tests were failing due to `reflect.Generic` no longer existing instead of failing due to inability to summon a mirror, dating back to when `Mirror` was called `Generic`. But I might have misunderstood.

File: library/src/scala/deriving/Mirror.scala (hunk 3 above)
Comment (dwijnand) [reply library/src/scala/deriving/Mirror.scala:61]: Plausibly because you could define elem types in terms of your own HList instead of the standard Tuple, e.g. shapeless might want that?  I have no strong feeling about that (and I don't know how active shapeless 3 is).

Comment (nicolasstucki) [reply library/src/scala/deriving/Mirror.scala:61]: @milessabin how is this used in shapeless 3?

Comment (nicolasstucki) [reply library/src/scala/deriving/Mirror.scala:61]: From what I can see shapeless uses Tuples.

Comment (nicolasstucki) [reply library/src/scala/deriving/Mirror.scala:61]: Actually here is an example that shows why it is not defined in `Mirror`: https://github.com/lampepfl/dotty/blob/master/tests/pos/mirror-implicit-scope.scala#L12

Comment (nicolasstucki) [reply library/src/scala/deriving/Mirror.scala:61]: The current version LGTM

**Comment by OlivierBlanvillain** [comment]
What are the benefits of adding these API to the standard library as opposed to defining then in user space? IIUC the Mirror API was not designed to provide a type safe experience out of the box, but rather to be a low-level building block for people to build type-safe frameworks on top...

**Comment by mpilquist** [comment]
Discoverability & avoidance of dependencies in foundational libraries, which lead to instability & churn in overall library ecosystem. The more type safety we can provide in these foundational APIs, the easier it will be for folks to build their higher level derivations, frameworks, etc.

**Comment by OlivierBlanvillain** [comment]
That makes sense, although if it's just an extension method away I think we could also do without. But I suppose such argument could be made for any addition to the standard library...

Regardless, I think that PR should also update `docs/docs/reference/contextual/derivation.md` since it's the main entry point to the Mirror APIs.

**Review by odersky — APPROVED** [review]
I think that addition is quite useful.

**Comment by odersky** [comment]
@mpilquist Can you rebase on latest master, then we can merge? Thank you!

**Comment by mpilquist** [comment]
Done, thanks!
